src/test/java/com/mmm/his/cer/utility/farser/ast_complex/ComplexFormulaAstTest.java

```diff
@@ -395,5 +395,35 @@ public void evaluateInTableFalse() throws Exception {
 
     assertThat(result.getResult(), is(false));
   }
+  
+  @Test
+  public void evaluateHasTrue() throws Exception {
+    String input = "a1bc HAS 1";
```

Comment:
Hmm, in which situations would the types be different? In this case, wouldn't they both be strings?

Or maybe in for something like "table has 1", where a number is needed?

Reply:
The use case I found for this is adding x days to a date. So on the left side, a LocalDate is returned, then on the right side an Integer is returned. In this example I added as a test they could easily both be strings, I just used an Integer as one to test it as different types.

There are other ways around this that does not involve opening this up to accept a left and right type that are different though, like using an Object as the left/right return type and then casting myself.

Reply:
Hmm... Thanks for the explanation. Internally to this library an `Object` type would be ok and pretty much the same, maybe even simpler than the need for using the "dummy" generic types and unchecked casting I believe. With this I mean not using generic types at all in the nodes.

Yes, for this edge case you _could_ use an `Object` type for these specific implementations.

For node implementations, the generic types are nice - but are they necessary? With generic types, a node implementation can only be implemented with that specific left/right/output type. With `Object` an implementation could accidentally use the wrong type for one of them which might only show up as runtime error. I could argue that its the implementers fault for not testing their node implementation well enough... :p

Any thoughts?

Reply:
I don't think the generic types are necessary, but they eliminate the need for implementations to deal with casting. If Object were used instead, then there would have to be some sort of conditional to see the type, then cast and perform the operation. The runtime error might actually happen less if Object were used because you'd be checking the type instead of assuming the type. I think for most cases, defining the generic type would result in a cleaner implementation though. @mfunaro what do you think?

Reply:
Sorry for the late reply. I think using Object definitely keeps thing simpler and forces the onus onto the implementation of the node and if the implementation doesn't cast correctly that is their issue to deal with. I am kind of leaning back toward using just Objects.